support/agent/miner_agent.go

```diff
@@ -0,0 +1,338 @@
+		ChainCommitRand:  []byte("not really random"),
+	}
+	return Message{
+		From:   ma.Worker,
+		To:     ma.RobustAddress,
+		Value:  big.Zero(),
+		Method: builtin.MethodsMiner.SubmitWindowedPoSt,
+		Params: &params,
+		ReturnHandler: func(v *vm.VM, _ Message, _ cbor.Marshaler) error {
+			return ma.scheduleNextProof(v, dlIdx)
+		},
+	}
+}
+
+// looks up sector deadline and partition so we can start adding it to PoSts
+func (ma *MinerAgent) registerSector(v *vm.VM, sectorNumber abi.SectorNumber) error {
+	var st miner.State
+	err := v.GetState(ma.IDAddress, &st)
+	if err != nil {
+		return err
+	}
+
+	dlIdx, pIdx, err := st.FindSector(v.Store(), sectorNumber)
+	if err != nil {
+		return err
+	}
+
+	if len(ma.deadlines[dlIdx]) == 0 {
+		err := ma.scheduleNextProof(v, dlIdx)
+		if err != nil {
+			return err
+		}
+	}
+
+	// pIdx should be sequential, but add empty partitions just in case
+	for pIdx >= uint64(len(ma.deadlines[dlIdx])) {
+		ma.deadlines[dlIdx] = append(ma.deadlines[dlIdx], bitfield.New())
+	}
+	ma.deadlines[dlIdx][pIdx].Set(uint64(sectorNumber))
+	return nil
+}
+
+// schedule a proof within the deadline's bounds
+func (ma *MinerAgent) scheduleNextProof(v *vm.VM, dlIdx uint64) error {
+	var st miner.State
+	err := v.GetState(ma.IDAddress, &st)
+	if err != nil {
+		return err
+	}
+
+	// find next proving window for this deadline
+	deadlineStart := st.ProvingPeriodStart + abi.ChainEpoch(dlIdx)*miner.WPoStChallengeWindow
+	if deadlineStart-miner.WPoStChallengeWindow < v.GetEpoch() {
+		deadlineStart += miner.WPoStProvingPeriod
+	}
+	deadlineClose := deadlineStart + miner.WPoStChallengeWindow
+	prooveAt := deadlineStart + abi.ChainEpoch(ma.rnd.Int63n(int64(deadlineClose-deadlineStart)))
+	ma.operationSchedule.ScheduleOp(prooveAt, proveDeadlineAction{dlIdx: dlIdx})
+	return nil
+}
+
+// create prove commit message
+func (ma *MinerAgent) createProveCommit(sectorNumber abi.SectorNumber) Message {
+	params := miner.ProveCommitSectorParams{
+		SectorNumber: sectorNumber,
+	}
+
+	return Message{
+		From:          ma.Worker,
+		To:            ma.RobustAddress,
+		Value:         big.Zero(),
+		Method:        builtin.MethodsMiner.ProveCommitSector,
+		Params:        &params,
+		ReturnHandler: ma.handleProveCommit,
+	}
+}
+
+// register an op for next epoch (after batch prove) to schedule a post for the sector
+func (ma *MinerAgent) handleProveCommit(v *vm.VM, msg Message, _ cbor.Marshaler) error {
+	sn := msg.Params.(*miner.ProveCommitSectorParams).SectorNumber
+	ma.PreCommits--
+	ma.operationSchedule.ScheduleOp(v.GetEpoch(), registerSectorAction{sectorNumber: sn})
+	return nil
+}
+
+// create precommit message and activation trigger
+func (ma *MinerAgent) createPrecommit(currentEpoch abi.ChainEpoch, sectorNumber abi.SectorNumber, sectorActivation abi.ChainEpoch) Message {
+	params := miner.PreCommitSectorParams{
+		SealProof:     ma.Config.ProofType,
+		SectorNumber:  sectorNumber,
+		SealedCID:     sectorSealCID(ma.rnd),
```

Comment:
My hope is to get more entropy into these simulations than we have in unit or scenario tests. Using seeded RNGs is just a quick way to get an even distribution. The downside is that small changes could completely alter the simulation (i.e. it's easy to reproduce problems but harder to confirm they're fixed).

One partial solution would be to used nested RNGs (a tree of RNG's where each node is seeded with a number taken from the parent). This would reduce but not eliminate the chance that a small change alters the sequence that leads to a problem.

Reply:
This was trivial to do per miner, so I went ahead and did it. Now changes in the random sequence of one miner will not affect the random sequences of other miners.